### PR TITLE
Replace TraceBuildState with Runtime Models

### PR DESCRIPTION
**Pull request by BradSwain**
The `TraceBuildState` only contain OpenMP logic (for now) and it is already becoming cluttered and leading to some bugs.

This PR replaces the "global" instance of state with separate runtime models. Logic for different runtimes (e.g. OpenMP, CUDA, etc.) can be split into their own class and interact with the trace construction in a more well defined way.

### CODE REVIEW

**Comment by BradSwain** [comment]
Only 1 test case is failing now, DRB106.

I checked the traces and it looks like the runtime model part is correct. Only 2 tasks are being created, and both by the master thread. Not sure why the race is not being reported.

File: src/Trace/Build/TraceBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+  // Else we reached the target instruction
+  // reset skipUntil ptr and continue traversing
+  state.skipUntil = nullptr;
+  return false;
+}
+
+// Check if currentThread is trying to create a recursive thread spawn at childEntry, by checking if the current thread
+// or any parent thread's entry functions are the same as childEntry's
+bool isRecursiveThreadSpawn(const ThreadTrace &currentThread, const pta::CallGraphNodeTy *childEntry) {
```
Comment (funemy) [inline src/Trace/Build/TraceBuilder.cpp:34]: Ideally we should allow the thread creation to recurse once then stop from spawning?

File: src/Trace/Build/TraceBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+      assert(entry && "Thread has no entry");
+
+      // Check for recursive thread creation
+      if (isRecursiveThreadSpawn(state.thread, entry)) {
+        llvm::outs() << "Skipping recursive thread creation: " << entry->getTargetFun()->getName() << "\n";
+        continue;
+      }
+      // Now we can push the event since we are sure we are going to crate a new thread
+      state.events.push_back(std::move(forkEventImpl));
+      // Note forkEventmpl has been moved and should not be accessed anymore
```
Comment (funemy) [inline src/Trace/Build/TraceBuilder.cpp:98]: change `Note` to `NOTE:`

**Review by funemy — CHANGES_REQUESTED** [review]
Just some minor changes

**Pull request merged**